src/QsCompiler/SyntaxProcessor/ExpressionVerification.fs

```diff
@@ -23,6 +25,28 @@ open Microsoft.Quantum.QsCompiler.Transformations.QsCodeOutput
 
 // utils for verifying types in expressions
 
+/// The context used for symbol resolution and type checking.
+type ResolutionContext<'a> =
+    { /// The symbol tracker for the parent callable.
+      Symbols : SymbolTracker<'a>
+      /// True if the parent callable for the current scope is an operation.
+      IsInOperation : bool
+      /// The return type of the parent callable for the current scope.
+      ReturnType : ResolvedType
+      /// The runtime capabilities for the compilation unit.
+      Capabilities : RuntimeCapabilities }
+
+/// <summary>Creates a resolution context for the specialization.</summary>
+/// <exception cref="Exception">Thrown if the specialization's parent does not exist.</exception>
+let CreateResolutionContext (nsManager : NamespaceManager) capabilities (spec : SpecializationDeclarationHeader) =
```

Comment:
I would go with a static member instead of just another function in the module: 

    static member Create (nsManager : NamespaceManager) capabilities (spec : SpecializationDeclarationHeader) : ResolutionContext<'a> =
        match nsManager.TryGetCallable spec.Parent (spec.Parent.Namespace, spec.SourceFile) with
        | Found declaration ->
            { Symbols = SymbolTracker (nsManager, spec.SourceFile, spec.Parent)
              IsInOperation = declaration.Kind = Operation
              ReturnType = StripPositionInfo.Apply declaration.Signature.ReturnType
              Capabilities = capabilities }
        | _ -> failwith "The specialization's parent does not exist."